File: Snooper/src/main/java/com/prateekj/snooper/formatter/XmlFormatter.java
package com.prateekj.snooper.formatter;

import android.util.Log;

import org.xml.sax.InputSource;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;

import javax.xml.transform.OutputKeys;
import javax.xml.transform.Source;
import javax.xml.transform.Transformer;
import javax.xml.transform.sax.SAXSource;
import javax.xml.transform.sax.SAXTransformerFactory;
import javax.xml.transform.stream.StreamResult;

public class XmlFormatter implements ResponseFormatter{

  public static final String TAG = XmlFormatter.class.getSimpleName();

  @Override
  public String format(String response) {
    try {
      Transformer serializer = SAXTransformerFactory.newInstance().newTransformer();
      serializer.setOutputProperty(OutputKeys.INDENT, "yes");
      serializer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
      serializer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2");
      Source xmlSource = new SAXSource(new InputSource(new ByteArrayInputStream(response.getBytes())));
      StreamResult res = new StreamResult(new ByteArrayOutputStream());
      serializer.transform(xmlSource, res);
      return new String(((ByteArrayOutputStream) res.getOutputStream()).toByteArray()).trim();
    } catch (Exception e) {
      e.printStackTrace();
      Log.d(TAG, e.getMessage(), e);
      return response;
    }
  }
}

File: Snooper/src/main/java/com/prateekj/snooper/formatter/XmlFormatter.kt
package com.prateekj.snooper.formatter

import com.prateekj.snooper.utils.Logger

import org.xml.sax.InputSource

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream

import javax.xml.transform.OutputKeys
import javax.xml.transform.sax.SAXSource
import javax.xml.transform.sax.SAXTransformerFactory
import javax.xml.transform.stream.StreamResult

class XmlFormatter : ResponseFormatter {

  override fun format(response: String): String {
    return try {
      val serializer = SAXTransformerFactory.newInstance().newTransformer().apply {
        setOutputProperty(OutputKeys.INDENT, "yes")
        setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes")
        setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2")
      }
      val xmlSource = SAXSource(InputSource(ByteArrayInputStream(response.toByteArray())))
      val res = StreamResult(ByteArrayOutputStream())
      serializer.transform(xmlSource, res)
      String((res.outputStream as ByteArrayOutputStream).toByteArray()).trim { it <= ' ' }
    } catch (e: Exception) {
      e.printStackTrace()
      Logger.e(TAG, e.message, e)
      response
    }

  }

  companion object {
    val TAG: String = XmlFormatter::class.java.simpleName
  }
}
